track2/src/icsi-scenic-tools-20120105/quicknet-v3_31/qncopy.man
. $Header: /u/drspeech/repos/quicknet2/qncopy.man,v 1.5 2007/02/19 09:03:32 davidj Exp $
.de EX		\" Begin Example
.ie \\n(.$ .nr EX \\$1n
.el \{\
.	ie n .nr EX 0n
.	el .nr EX 0n
.\}
.in +\\n(EXu
.if n .sp 1
.if t .sp .5v
.nf
.CW
.ft CB
..
.de EE		\" End example
.in -\\n(EXu
.fi
.}f             \" restore basic text attributes
.if n .sp 1
.if t .sp .5v
..
.de MS
.ie t \&\f(CB\\$1\f(NR(\\$2)\\$3
.el \fI\\$1\fP(\\$2)\\$3
..
.TH qncopy 1 "$Date: 2007/02/19 09:03:32 $" ICSI "ICSI SPEECH SOFTWARE"
.SH NAME
qncopy \- Copy QuickNet data files, possibly with modifications
.SH SYNOPSIS
.B qncopy
[
.I options
]
.SH DESCRIPTION
.I qncopy
is a simple program for copying one speech feature archive 
to another.  It is really just a wrapper around the QuickNet 
"feature stream" classes, to allow access to some of the 
enhanced functions they provide.  It mirrors many of the 
feature processing options provided by 
.MS qnstrn 1
and
.MS qnsfwd 1 .
.SH OPTIONS
.I qncopy
can take the following options (defaults in parens):
.TP 1i			\" Indent the paragraph
.BI ftr_file= filename (required)
Specify the input speech feature archive file.
.TP
\fBftr_format\fR=\fItype\fR ("pfile")
The format of the input feature file.  Supported formats include 
"pfile" (see
.MS pfile 5 )
"pre" and "lna" (see
.MS lna 5 ).
.TP
\fBftr_width\fR=\fIwidth\fR (0)
The number of feature columns in the feature file.  This can be 
left at zero for pfiles, which specify their feature width in 
the header, but must be specified for pre and lna files, which 
do not.
.TP
\fBftr_norm_file\fR=\fIfilename\fR 
If specified, means and standard deviations for each feature column 
(including any that are calculated on-line) are read from this file 
and applied as normalization.  See \f4norm_mode\f1 below.
.TP
\fBoutftr_file\fR=\fIfilename\fR ("-")
The file that is written by the program.  "-" means to write to 
standard output.
.TP
.BI delta_order= order (0)
The highest order of delta features to calculate.  0 means that no 
delta caculation is performed.  1 means to calculate just the 1st order
derivatives.  2 means to calculate both first and second order (curvature 
or acceleration) derivatives.
Note that the filters are not scaled by the sum of squares, so
mean-variance normalization is recommended.
.TP
.BI delta_win= size (9)
The size of the window used to calculate first- and second-order derivatives, 
in samples.
.TP
.BI ftr_start= num (0)
Of the internally-constructed feature stream (i.e. including any deltas), 
this is the first column to be written.
.TP
.BI ftr_count= num (0)
The total number of features from the internal feature stream which 
are written.
.TP
.BI sent_start= integer (0)
The number of the first sentence in the feature file to use
(0 means the first).
.TP
.BI sent_count= integer
The total number of sentences in the feature file to use.  Specifying
more than the available number of sentences will result in all remaining 
sentences being used, which is the default.
.TP
.BI sent_range= range-spec-string
The sentences to process, specified by a Range token in 
one of the formats defined by 
.MS QN_Range 3 .  
This is an alternative 
to using sent_start and sent_count.  
.TP
\fBnorm_mode\fR=\fImode\fR ("file")
Feature normalization mode.  The default of \f4file\f1 means to use global 
normalization constants, read from \f4ftr_norm_file\f1, for 
the entire data set.  (If \f4ftr_norm_file\f1 is empty, no 
normalization is performed).  \f4utts\f1 means to normalize each utterance 
individually, similar to 
.MS pfile_normutts 1 ,
such that within each utterance each 
feature dimension is separately scaled and shifted to have zero 
mean and unit variance (this is actually accomplished, 
through the QN_InFtrStream_NormUtts class, by reading through 
each utterance twice - once to gather the normalization constants, 
the second time to actually process the features).  \f4online\f1 
means to apply online normalization, initializing at the start of 
each utterance to the values read from \f4ftr_norm_file\f1, but 
then adapting with smoothed estimates of the means and variances 
of each feature dimension, according to the smoothing constants 
\f4norm_alpha_m\f1 and \f4norm_alpha_v\f1.
.TP
\fBnorm_alpha_m\fR=\fIval\fR (0.005)
Update constant for the recursive mean estimates when \f4norm_mode=onl\f1.
At each step, the mean estimate is update to (1-alpha)*oldmean + alpha*x.
.TP
\fBnorm_alpha_v\fR=\fIval\fR (0.005)
Update constant for the recursive variance estimates when \f4norm_mode=onl\f1.
.TP
\fBindexedop\fR=\fIbool\fR (1)
Controls whether the output file is created in 'indexed' mode i.e. a 
random-access index will be written to the end of the output file 
when it is closed.  (I think this is only supported for pfiles).
.TP
.BI debug= integer
Set the level of debugging output.  \fB0\fR means none, \fB6\fR means lots!
.TP
.BI verbose= bool
Output more log information if
.BR true .
.SH BUGS
Doesn't support label fields or pure-label pfiles.
.SH AUTHOR
Dan Ellis  <dpwe@icsi.berkeley.edu>, making extensive use of the 
original QuickNet classes by David Johnson <davidj@icsi.berkeley.edu>
.SH SEE ALSO
.MS pfile_normutts 1 ,
.MS feacat 1 ,
.MS lna 5 ,
.MS pfile 5 .


track2/src/icsi-scenic-tools-20120105/quicknet-v3_31/qncopy.cc
#ifndef NO_RCSID
const char* qncopy_rcsid = 
    "$Header: /u/drspeech/repos/quicknet2/qncopy.cc,v 1.11 2010/10/29 02:20:38 davidj Exp $";
#endif

/* Must include the config.h file first */
#include <QN_config.h>
#include <float.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#ifdef QN_HAVE_LIMITS_H
#include <limits.h>
#endif
#ifndef EXIT_SUCCESS
#define EXIT_SUCCESS (0)
#define EXIT_FAILURE (1)
#endif
#include <sys/types.h>
#ifdef QN_HAVE_SYS_TIME_H
#include <sys/time.h>
#endif
#ifdef QN_HAVE_SYS_PARAM_H
#include <sys/param.h>
#endif
#include <unistd.h>


#ifndef FILENAME_MAX
#define FILENAME_MAX (MAXPATHLEN)
#endif

#include "QuickNet.h"

static struct {
    const char* ftr_filename;
    const char* ftr_format;
    int ftr_width;
    const char* ftr_norm_file;
    const char* outftr_filename;
    int delta_win;		// length of delta window
    int delta_order;		// order of deltas to add
    int first_ftr;		// first feature to filter
    int num_ftrs;		// how many features to filter
    int first_sent;		// First sentence to process
    int num_sents;		// Number of sentences to process
    const char *sent_range;	        // arbitrary sentence specification
    const char *norm_mode_str;	// token from command line for norm_mode
    int norm_mode;		// Normalize from file, per utt, or online?
    double norm_am;		// time const for onlnorm means
    double norm_av;		// time const for onlnorm vars
    int indexedop;		// Build an index in the output file?
    int dbg;			// Debug level.
    int verbose;
} config;

void
set_defaults(void)
{
    config.ftr_filename = NULL;
    config.ftr_format = "pfile";
    config.ftr_width = 0;
    config.ftr_norm_file = NULL;
    config.outftr_filename = "-";
    config.delta_win = 9;
    config.delta_order = 0;	// by default, don't add deltas
    config.first_ftr = 0;
    config.num_ftrs= 0;		// default to all ftrs if not spec'd
    config.first_sent = 0;	// First sentence to recognize.
    config.num_sents = INT_MAX;	// Number of sentences to recognize.
    config.sent_range = 0;
    config.norm_mode_str = NULL;
    config.norm_mode = QN_NORM_FILE;
    config.norm_am = QN_DFLT_NORM_AM;
    config.norm_av = QN_DFLT_NORM_AV;
    config.indexedop = 1;
    config.dbg = 0;
    config.verbose = 1;
}

QN_ArgEntry argtab[] =
{
{ NULL, "QuickNet speech feature file copyer version " QN_VERSION, QN_ARG_DESC },
{ "ftr_file", "Input feature file", QN_ARG_STR,
  &(config.ftr_filename), QN_ARG_REQ },
{ "ftr_format", "Input file format", QN_ARG_STR, 
  &(config.ftr_format) }, 
{ "ftr_width", "Input file width (for pre, lna)", QN_ARG_INT, 
  &(config.ftr_width) },
{ "ftr_norm_file", "Normalize features from norms file", QN_ARG_STR, 
  &(config.ftr_norm_file) },
{ "outftr_file", "Output feature file", QN_ARG_STR,
  &(config.outftr_filename) },
{ "delta_win", "Length of delta-filter", QN_ARG_INT, 
  &(config.delta_win) },
{ "delta_order", "Add delta features to what degree?", QN_ARG_INT, 
  &(config.delta_order) },
{ "ftr_start", "First feature column (of net stream) to write", 
  QN_ARG_INT, &(config.first_ftr) },
{ "ftr_count", "Number of feature columns to write",
  QN_ARG_INT, &(config.num_ftrs) },
{ "sent_start", "Number of first sentence",
  QN_ARG_INT, &(config.first_sent) },
{ "sent_count", "Number of sentences",
  QN_ARG_INT, &(config.num_sents) },
{ "sent_range", "Specify sentences to use as a QN_Range(3) token", 
  QN_ARG_STR, &(config.sent_range) },
{ "norm_mode", "Mode for normalization - file/utt/online", 
  QN_ARG_STR, &(config.norm_mode_str) },
{ "norm_alpha_m", "Update constant for online norm means", 
  QN_ARG_DOUBLE, &(config.norm_am) },
{ "norm_alpha_v", "Update constant for online norm vars", 
  QN_ARG_DOUBLE, &(config.norm_av) },
{ "indexedop", "Build an index for the output file?",
  QN_ARG_INT, &(config.indexedop) },
{ "debug", "Output additional diagnostic information",
  QN_ARG_INT, &(config.dbg) },
{ "verbose", "Output diagnostic information",
  QN_ARG_BOOL, &(config.verbose) },
{ NULL, NULL, QN_ARG_NOMOREARGS }
};


static void
copy_run(FILE* out_file, FILE* logfile, int verbose)
{
    if (verbose>0) {
	// A note for the logfile.
        time_t now = time(NULL);
	fprintf(logfile, "\nProgram version: qncopy %s.\n", QN_VERSION);
#ifdef QN_HAVE_GETHOSTNAME
	char hostname[MAXHOSTNAMELEN];
	assert(gethostname(hostname, MAXHOSTNAMELEN)==0);
	fprintf(logfile, "Host: %s\n", hostname);
#endif
	fprintf(logfile, "Program start: %s", ctime(&now));
    }
    if (verbose>0) {
	fprintf(logfile, "Opening feature file...\n");
    }

    int indexed = (config.sent_range==0)?0:1;	// has to be indexed to cut
    int buffer_frames = 500;
    size_t num_sents;
    // Command line arguments use INT_MAX for "all sentences", streams
    // use QN_ALL, which has a different value.
    if (config.num_sents==INT_MAX) {
	num_sents = QN_ALL;
    } else {
	num_sents = config.num_sents;
    }
    FILE *norm_file = NULL;
    if (config.ftr_norm_file != NULL) {
	norm_file = fopen(config.ftr_norm_file, "r");
	if (norm_file == NULL) {
	    QN_ERROR(NULL, "failed to open norm file %s for reading.",
		     config.ftr_norm_file);
	}
    }    

    QN_InFtrStream* ftr_str = QN_build_ftrstream(config.dbg, "ftrfile", 
				config.ftr_filename, config.ftr_format, 
				config.ftr_width, indexed, 
				norm_file, 
				config.first_ftr, config.num_ftrs, 
				config.first_sent, num_sents, buffer_frames,
				config.delta_order, config.delta_win, 
				config.norm_mode, config.norm_am, 
				config.norm_av);

    if (norm_file)	fclose(norm_file);

    // Select sentences we need.
    if (config.sent_range) {
	QN_InFtrStream_CutRange* fwd_ftr_str = 
	    new QN_InFtrStream_CutRange(config.dbg, "ftrfile",
					*ftr_str, config.sent_range);
	//	QN_InFtrStream_Cut* fwd_ftr_str = 
	//            new QN_InFtrStream_Cut(config.dbg, "ftrfile",
	//				  *ftr_str, 1, 10);
	ftr_str = (QN_InFtrStream*)fwd_ftr_str;
    }


    const size_t n_ftrs = ftr_str->num_ftrs();

    if (verbose>0) {
	fprintf(logfile, "Data will have %lu feature(s).\n",
		(unsigned long) n_ftrs);
        fprintf(logfile, "Copying features...\n");
    }

    int buffrms = 1000;
    float *ftrbuf = new float[buffrms * n_ftrs];

    int tot_utts = 0;
    int tot_frms = 0;

    // Create the output stream
    int n_labs = 0;	// cannot copy labels to output - so write none
    QN_OutFtrStream *outpf;
    outpf = new QN_OutFtrLabStream_PFile(config.dbg, "outpfile", out_file, 
					 n_ftrs, n_labs, config.indexedop);
    QN_SegID segid;
    while( (segid = ftr_str->nextseg()) != QN_SEGID_BAD) {
	size_t nred;
	while( (nred =ftr_str->read_ftrs(buffrms, ftrbuf)) > 0) {
	    outpf->write_ftrs(nred, ftrbuf);
	    tot_frms += nred;
	}
	outpf->doneseg(segid);
	++tot_utts;
    }

    // Close the output file
    delete outpf;

    delete ftr_str;

    if (verbose) {
	fprintf(logfile, "%d utts (%d frames) written: %s\n", 
		tot_utts, tot_frms, config.outftr_filename);
    }
    if (verbose) {
	// A note for the logfile.
        time_t now = time(NULL);
	fprintf(logfile, "Program stop: %s", ctime(&now));
    }
}

int
main(int argc, const char* argv[])
{
    char* progname;		// The name of the program - set by QN_args.
    FILE* out;
    FILE* logfile;
    char logbuf[160];

    set_defaults();
    QN_initargs(&argtab[0], &argc, &argv, &progname);

    // map norm_mode_str to val
    config.norm_mode = QN_string_to_norm_const(config.norm_mode_str);

    // Check argument sanity
    if (config.delta_order > 0) {
	if (config.delta_order > 2) {
	    fprintf(stderr, "qncopy: delta_order %d must be between 0 and 2\n", 
		    config.delta_order);
	    exit(-1);
	}
	if (config.delta_win < 3 || (config.delta_win & 1) == 0) {
	    fprintf(stderr, "qncopy: delta_window %d must be odd and > 1\n",
		    config.delta_win);
	    exit(-1);
	}
    }

    logfile = stderr;

    // Enable line buffering on log file.
    // *** Note: there is a bug with newlib-1.6.1 (as used on SPERT) that
    // *** makes setvbuf fail if the second argument is NULL.
    if (setvbuf(logfile, logbuf, _IOLBF, sizeof(logbuf)))
	abort();		// Fail on error
    if (config.verbose) {
	QN_printargs(logfile, progname, &argtab[0]);
	QN_logger = new QN_Logger_Simple(logfile, stderr, progname);
    } else {
	logfile = NULL;
	QN_logger = new QN_Logger_Simple(logfile, stderr, progname);
    }
    // Ensure the lines are sent to the logfile as soon as they are printed.

    out = fopen(config.outftr_filename, "w");
    if (out==NULL) {
	QN_ERROR(NULL, "failed to open ftr file %s for writing.",
		 config.outftr_filename);
    }

    // Do the transfer
    copy_run(out, logfile, config.verbose);

    if (fclose(out)) {
	QN_ERROR(NULL, "failed to close output file %s.",
		 config.outftr_filename);
    }
    QN_close_ftrfiles();

    exit(EXIT_SUCCESS);
}
